adr/ADR-42.md

```diff
@@ -0,0 +1,166 @@
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+Here we state that the Consumer has one group called `jobs`, it is operating on the `overflow` policy and requires `explicit`
+Acks, any other Ack policy will produce an error. If we force this ack policy in normal use we should error in Pedantic mode.
+
+Pull requests will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"min_pending": 1000` - only deliver messages when `num_pending` for the consumer is >= 1000
+ * `"min_ack_pending: 1000` - only deliver messages when `ack_pending` for the consumer is >= 1000
+
+If `min_pending` and `min_ack_pending` are both given either being satisfied will result in delivery (boolean OR).
+
+In the specific case where MaxAckPending is 1 and a pull is made using `min_pending: 1` this should only be served when
+there are no other pulls waiting. This means we have to give priority to pulls without conditions over those with when
+considering the next pull that will receive a message.
+
+### `pinned_client` policy
+
+Users want to have a single client perform all the work for a consumer, but they also want to have a stand-by client that
+can take over when the primary, aka `pinned` client, fails.
+
+**NOTE: We should not describe this in terms of exclusivity as there is no such guarantee, there will be times when one
+client think it is pinned when it is not because the server switched.** 
+
+The `pinned_client` policy provides server-side orchestration for the selection of the pinned client.
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "pinned_client",
+	PriorityTimeout: 120*time.Second,
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+This configuration states:
+
+ * We have 1 group defined and all pulls have to belong to this group
+ * The policy is `pinned_client` that activates these behaviors
+ * When a pinned client has not done any pulls in the last 120 seconds the server will switch to another client
+ * AckPolicy has to be `explicit`. If we force this ack policy in normal use we should error in Pedantic mode
+
+A pull request will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"id": "xyz"` - the pinned client will have this ID set to the one the server last supplied (see below), otherwise 
+   this field is absent
+
+After selecting a new pinned client, the first message that will be delivered to this client, and all future ones, will 
```

Comment:
There are two things to discuss here:
1. This will be the first time we send status/api-like header on an actual user message. Nothing bed in it. Just mentioning that this is a new pattern.
All other headers set in msgs by the server are informational, not status-like.
2. Sending the header in every message will have some overhead. I wonder if its not enough to sent the initial message. After all, sending the wrong along the pull request will trigger a server error message.

Reply:
Regarding when we send the header, I did that to guard against the first message being a magical message that can never be lost which would lead to a reselection happening after timeout. While if its on each message the first can be lost and things will still work - that first will just later be retried.

Seems worth it to me, wdyt @Jarema

Reply:
Yes, I agree with that. However I still do not like the overhead. What about sending the `Nats-Pin-Id` in each message in first batch? 
Missing whole batch might suggest anyway a problem with a client. But yeah, there is also a batch of one...

Ok, let's stich to sending it in every message.
We can always revisit.